File: apps/mobile/lib/shared/auth/permissions_service.dart
import 'app_permissions.dart';
import '../../api/api_client.dart';

class PermissionsService {
  PermissionsService(this._apiClient);

  final ApiClient _apiClient;

  Future<AppPermissions> getPermissions() async {
    try {
      final data = await _apiClient.getJsonFlexible(
        '/permissions',
      );
      final payload = _extractPayload(data);
      final canCreate = payload['canCreate'];
      final createPermissions = _parseCreatePermissions(canCreate);
      return AppPermissions(
        role: payload['role']?.toString() ?? 'USER',
        isAdmin: _readBool(payload['isAdmin']) ?? false,
        canCreate: createPermissions,
        canModerateUserContent:
            _readBool(payload['canModerateUserContent']) ?? false,
        canManageResidents: _readBool(payload['canManageResidents']) ?? false,
        canGenerateActivationCodes:
            _readBool(payload['canGenerateActivationCodes']) ?? false,
        canManageRoles: _readBool(payload['canManageRoles']) ?? false,
      );
    } catch (_) {
      return AppPermissions.empty;
    }
  }

  AppPermissions _mapPermissions(Map<String, dynamic> payload) {
    final isAdmin = _readBool(payload['isAdmin']) ?? false;
    final legacyManage = _readBool(payload['canManageContent']) ??
        _readBool(payload['admin']) ??
        _readBool(payload['canManage']);
    final canCreateEvents =
        _readBool(payload['canCreateEvents']) ?? legacyManage;
    final canCreatePosts =
        _readBool(payload['canCreatePosts']) ?? legacyManage;
    final canCreateNews = _readBool(payload['canCreateNews']) ?? legacyManage;
    final canCreateWarnings =
        _readBool(payload['canCreateWarnings']) ?? legacyManage;
    final canModerate = _readBool(payload['canModerate']) ?? legacyManage;
    final canManageResidents = _readBool(payload['canManageResidents']);
    final canGenerateActivationCodes =
        _readBool(payload['canGenerateActivationCodes']);

    final roleValue = payload['role'];
    final parsedRole = _parseRole(roleValue);
    final inferredRole = parsedRole ?? _inferRole(
      isAdmin: isAdmin,
      canCreateEvents: canCreateEvents,
      canCreatePosts: canCreatePosts,
      canCreateNews: canCreateNews,
      canCreateWarnings: canCreateWarnings,
      canModerate: canModerate,
    );

    return AppPermissions(
      role: inferredRole,
      isAdmin: isAdmin,
      canCreateEvents: canCreateEvents ?? false,
      canCreatePosts: canCreatePosts ?? false,
      canCreateNews: canCreateNews ?? false,
      canCreateWarnings: canCreateWarnings ?? false,
      canModerate: canModerate ?? false,
      canManageResidents: canManageResidents ?? false,
      canGenerateActivationCodes: canGenerateActivationCodes ?? false,
    );
  }

  AppRole? _parseRole(dynamic value) {
    if (value is! String) {
      return null;
    }
    switch (value.toUpperCase()) {
      case 'ADMIN':
        return AppRole.admin;
      case 'STAFF':
        return AppRole.staff;
      case 'USER':
        return AppRole.user;
      default:
        return null;
    }
  }

  AppRole _inferRole({
    required bool isAdmin,
    required bool? canCreateEvents,
    required bool? canCreatePosts,
    required bool? canCreateNews,
    required bool? canCreateWarnings,
    required bool? canModerate,
  }) {
    if (isAdmin) {
      return AppRole.admin;
    }
    final hasContentPermission =
        (canCreateEvents ?? false) ||
            (canCreatePosts ?? false) ||
            (canCreateNews ?? false) ||
            (canCreateWarnings ?? false) ||
            (canModerate ?? false);
    return hasContentPermission ? AppRole.staff : AppRole.user;
  }

  Map<String, dynamic> _extractPayload(dynamic data) {
    if (data is Map<String, dynamic>) {
      final nested = data['data'] ?? data['permissions'];
      if (nested is Map<String, dynamic>) {
        return nested;
      }
      return data;
    }
    return const <String, dynamic>{};
  }

  bool? _readBool(dynamic value) {
    if (value is bool) {
      return value;
    }
    if (value is String) {
      if (value.toLowerCase() == 'true') {
        return true;
      }
      if (value.toLowerCase() == 'false') {
        return false;
      }
    }
    if (value is num) {
      return value != 0;
    }
    return null;
  }

  CreatePermissions _parseCreatePermissions(dynamic value) {
    if (value is Map<String, dynamic>) {
      return CreatePermissions(
        marketplace: _readBool(value['marketplace']) ?? false,
        help: _readBool(value['help']) ?? false,
        movingClearance: _readBool(value['movingClearance']) ?? false,
        cafeMeetup: _readBool(value['cafeMeetup']) ?? false,
        kidsMeetup: _readBool(value['kidsMeetup']) ?? false,
        apartmentSearch: _readBool(value['apartmentSearch']) ?? false,
        lostFound: _readBool(value['lostFound']) ?? false,
        rideSharing: _readBool(value['rideSharing']) ?? false,
        jobsLocal: _readBool(value['jobsLocal']) ?? false,
        volunteering: _readBool(value['volunteering']) ?? false,
        giveaway: _readBool(value['giveaway']) ?? false,
        skillExchange: _readBool(value['skillExchange']) ?? false,
        officialNews: _readBool(value['officialNews']) ?? false,
        officialWarnings: _readBool(value['officialWarnings']) ?? false,
        officialEvents: _readBool(value['officialEvents']) ?? false,
      );
    }
    return CreatePermissions.empty;
  }
}

File: apps/api/src/permissions/permissions.service.ts
import { Injectable } from '@nestjs/common';
import { ContentType } from '../content/content.types';
import { verifyAccessToken } from '../auth/jwt.utils';
import { UserRole } from '../auth/user-roles';

type PermissionsPayload = {
  role: UserRole;
  isAdmin: boolean;
  canCreate: {
    marketplace: boolean;
    help: boolean;
    movingClearance: boolean;
    cafeMeetup: boolean;
    kidsMeetup: boolean;
    apartmentSearch: boolean;
    lostFound: boolean;
    rideSharing: boolean;
    jobsLocal: boolean;
    volunteering: boolean;
    giveaway: boolean;
    skillExchange: boolean;
    officialNews: boolean;
    officialWarnings: boolean;
    officialEvents: boolean;
  };
  canModerateUserContent: boolean;
  canManageResidents: boolean;
  canGenerateActivationCodes: boolean;
  canManageRoles: boolean;
};

const userCreatePermissions = {
  marketplace: true,
  help: true,
  movingClearance: true,
  cafeMeetup: true,
  kidsMeetup: true,
  apartmentSearch: true,
  lostFound: true,
  rideSharing: true,
  jobsLocal: true,
  volunteering: true,
  giveaway: true,
  skillExchange: true,
};

const touristCreatePermissions = {
  marketplace: false,
  help: false,
  movingClearance: false,
  cafeMeetup: false,
  kidsMeetup: false,
  apartmentSearch: false,
  lostFound: false,
  rideSharing: false,
  jobsLocal: false,
  volunteering: false,
  giveaway: false,
  skillExchange: false,
};

@Injectable()
export class PermissionsService {
  getPermissions(
    headers: Record<string, string | string[] | undefined>,
  ): PermissionsPayload {
    const payload = verifyAccessToken(headers);
    const role = payload?.role ?? UserRole.USER;

    const isTourist = role === UserRole.TOURIST;
    const canCreateOfficial =
      role === UserRole.STAFF || role === UserRole.ADMIN;
    const canModerateUserContent =
      role === UserRole.STAFF || role === UserRole.ADMIN;
    const isAdmin = role === UserRole.ADMIN;
    const canCreateUserContent = isTourist
      ? touristCreatePermissions
      : userCreatePermissions;

    return {
      role,
      isAdmin,
      canCreate: {
        ...canCreateUserContent,
        officialNews: canCreateOfficial,
        officialWarnings: canCreateOfficial,
        officialEvents: canCreateOfficial,
      },
      canModerateUserContent: isTourist ? false : canModerateUserContent,
      canManageResidents: isAdmin,
      canGenerateActivationCodes: isAdmin,
      canManageRoles: isAdmin,
    };
  }

  canCreateOfficialContent(role: UserRole) {
    return role === UserRole.STAFF || role === UserRole.ADMIN;
  }

  isUserContent(type: ContentType) {
    return (
      type !== ContentType.OFFICIAL_EVENT &&
      type !== ContentType.OFFICIAL_NEWS &&
      type !== ContentType.OFFICIAL_WARNING
    );
  }
}
